hardware/raspberrypi/bootmodes/bootflow_2711.md

```diff
@@ -1,31 +1,74 @@
+* If nRPIBOOT GPIO is high or OTP does NOT define `nRPIBOOT` GPIO 
+   * Check OTP to see if recovery.bin can be loaded from SD/EMMC
+      * If SD recovery.bin is enabled then check primary SD/EMMC for `recovery.bin`
+         * Success - run `recovery.bin`
+         * Fail - continue
+   * Check SPI EEPROM for second stage loader
+      * Success - run second stage bootloader 
+      * Fail - continue
+* While True
+   * Attempt to load recovery.bin from [USB device boot](../computemodule/cm-emmc-flashing.md)
+      * Success - run `recovery.bin`
+      * Fail - retry USB device boot
 
+N.B. Currently only CM4 reserves a GPIO for `nRPIBOOT`
 
-## SD Card Boot
-The bootloader loads the files in the [boot folder](../../../configuration/boot_folder.md) according to the [boot options](../../../configuration/config-txt/boot.md) in config.txt
+## recovery.bin
+`recovery.bin` is a minimal second stage program used to reflash the bootloader SPI EEPROM image.
 
-## Network boot
+# Second stage bootloader 
 
-Details of the network booting can be found [here](../bcm2711_bootloader_config.md)
+This section describes the high-level flow of the second stage bootloader.
 
-## USB mass storage boot
-
-USB booting is still under development.
-
-
-## BOOT_ORDER
-
-The `BOOT_ORDER` configuration item is embedded inside the bootloader code in the EEPROM. See the [Pi4 Bootloader Configuration](../bcm2711_bootloader_config.md) page for details on how to change the bootloader configuration.
+Please see the [bootloader configuration](../bcm2711_bootloader_config.md) page for more information about each boot-mode and the [boot folder](../../../configuration/boot_folder.md) page for a description of the GPU firmware files loaded by this stage.
 
+* Initialise clocks and SDRAM
+* Read the EEPROM configuration file
+* Check `PM_RSTS` register to determine if HALT is requested
+   * Check `POWER_OFF_ON_HALT` and `WAKE_ON_GPIO` EEPROM configuration settings.
+   * If `POWER_OFF_ON_HALT` is `1` and `WAKE_ON_GPIO` is `0` then
+      * Use PMIC to power off system
+   * else
+      * While GPIO3 is high OR `WAKE_ON_GPIO` is 1
+         Sleep
+* While True
+   * Read the next boot-mode from the BOOT_ORDER parameter in the EEPROM config file.
+   * If boot-mode == `RESTART`
+      * Jump back to the first boot-mode in the `BOOT_ORDER` field
+   * else if boot-mode == `STOP`
+      * Display start.elf not found [error pattern](../../../configuration/led_blink_warnings.md) and wait forever.
+   * else if boot-mode == `SD CARD`
+      * Attempt to load firmware from the SD card
+         * Success - run the firmware
+         * Failure - continue   
+   * If boot-mode == `NETWORK` then 
+      * Use DHCP protocol to request IP address
+      * Load firmware from the DHCP or statically defined TFTP server
+      * If the firmware is not found or a timeout or network error occurs then continue
+   * else if boot-mode == `USB-MSD` or boot-mode == `USB-BCM-MSD` then
```

Comment:
https://github.com/raspberrypi/documentation/blob/master/hardware/raspberrypi/bcm2711_bootloader_config.md#boot_order says `BCM-USB-MSD` rather than `USB-BCM-MSD`